File: core-test/src/main/java/com/github/simonpercic/oklog/core/TestUtils.java
package com.github.simonpercic.oklog.core;

import org.mockito.ArgumentCaptor;

import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.verify;

/**
 * @author Simon Percic <a href="https://github.com/simonpercic">https://github.com/simonpercic</a>
 */
public final class TestUtils {

    private TestUtils() {
        // no instance
    }

    public static LogDataBuilder getLogData(LogManager logManager) {
        ArgumentCaptor<LogDataBuilder> dataCaptor = ArgumentCaptor.forClass(LogDataBuilder.class);
        verify(logManager).log(dataCaptor.capture());
        return dataCaptor.getValue();
    }

    public static void assertData(LogDataBuilder expected, LogDataBuilder value) {
        assertEquals(expected.getRequestMethod(), value.getRequestMethod());
        assertEquals(expected.getRequestUrl(), value.getRequestUrl());
        assertEquals(expected.getRequestContentType(), value.getRequestContentType());
        assertEquals(expected.getRequestContentLength(), value.getRequestContentLength());
        assertEquals(expected.getRequestBody(), value.getRequestBody());
        assertEquals(expected.getRequestBodyState(), value.getRequestBodyState());
        assertEquals(expected.isRequestFailed(), value.isRequestFailed());
        assertEquals(expected.getResponseCode(), value.getResponseCode());
        assertEquals(expected.getResponseMessage(), value.getResponseMessage());
        assertEquals(expected.getResponseUrl(), value.getResponseUrl());
        assertEquals(expected.getResponseContentLength(), value.getResponseContentLength());
        assertEquals(expected.getResponseBodyState(), value.getResponseBodyState());
        assertEquals(expected.getResponseBodySize(), value.getResponseBodySize());
        assertEquals(expected.getResponseBody(), value.getResponseBody());
    }
}

File: core-test/src/main/kotlin/com/github/simonpercic/oklog/core/TestUtils.kt
package com.github.simonpercic.oklog.core

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.mockito.ArgumentCaptor
import org.mockito.Mockito.verify

/**
 * @author Simon Percic [https://github.com/simonpercic](https://github.com/simonpercic)
 */
object TestUtils {

    fun getLogData(logManager: LogManager): LogDataBuilder {
        val dataCaptor = ArgumentCaptor.forClass(LogDataBuilder::class.java)
        verify(logManager).log(dataCaptor.capture())
        return dataCaptor.value
    }

    internal fun assertData(expected: LogDataBuilder, value: LogDataBuilder) {
        assertEquals(expected.requestMethod, value.requestMethod)
        assertEquals(expected.requestUrl, value.requestUrl)
        assertEquals(expected.requestUrlPath, value.requestUrlPath)
        assertEquals(expected.requestContentType, value.requestContentType)
        assertEquals(expected.requestContentLength, value.requestContentLength)
        assertEquals(expected.requestBody, value.requestBody)
        assertEquals(expected.requestBodyState, value.requestBodyState)
        assertEquals(expected.isRequestFailed, value.isRequestFailed)
        assertEquals(expected.responseCode.toLong(), value.responseCode.toLong())
        assertEquals(expected.responseMessage, value.responseMessage)
        assertEquals(expected.responseUrl, value.responseUrl)
        assertEquals(expected.responseContentLength, value.responseContentLength)
        assertEquals(expected.responseBodyState, value.responseBodyState)
        assertEquals(expected.responseBodySize, value.responseBodySize)
        assertEquals(expected.responseBody, value.responseBody)
    }

    fun assertData(requestMethod: String, requestUrl: String, requestUrlPath: String,
        requestBodyState: LogDataBuilder.BodyState, failed: Boolean, value: LogDataBuilder) {

        val expectedValue = LogDataBuilder()
            .requestMethod(requestMethod)
            .requestUrl(requestUrl)
            .requestUrlPath(requestUrlPath)
            .requestBodyState(requestBodyState)

        if (failed) {
            expectedValue.requestFailed()
        }

        assertData(expectedValue, value)
    }

    fun assertNoRequestHeaders(value: LogDataBuilder) {
        assertNull(value.requestHeaders)
    }
}
